srk/src/terminationLLRF.ml

```diff
@@ -0,0 +1,425 @@
+            BatList.map 
+              (fun (generator, _) ->
+                 let pre_trans_term = get_orig_or_primed_expr_of_gen generator x_list in
+                 let post_trans_term = get_orig_or_primed_expr_of_gen generator xp_list in
+                 let equ = mk_eq srk post_trans_term pre_trans_term in
+                 equ
+              )
+              gens
+          in
+          let restricted_formula = mk_and srk (f :: new_constraints) in
+          logf "\nrestricted formula for next iter:\n%s\n\n" (Formula.show srk restricted_formula);
+          match Smt.equiv srk formula restricted_formula with
+          | `Unknown -> logf ~attributes:[`Bold; `Red] "Cannot decide if there is any improvement at this level"; (false, depth, formula, qrfs)
+          | `Yes -> logf ~attributes:[`Bold; `Red] "No improvement at this level, halt LLRF synthesis"; (false, depth, formula, qrfs)
+          | `No -> logf ~attributes:[`Bold; `Green] "There is improvement at this level";
+              match Smt.get_model srk restricted_formula with
+              | `Sat _ -> 
+                logf ~attributes:[`Bold; `Yellow] "\n\n\nTransition formula SAT, try to synthesize next depth\n\n";
+                find_quasi_rf (depth+1) srk restricted_formula new_qrfs x_list xp_list dx_list x_set xp_set dx_set x_to_dx dx_to_x coeff_x_list coeff_x_set
+              | `Unknown -> failwith "SMT solver should not return unknown"
+              | `Unsat -> (logf ~attributes:[`Bold; `Green] "Transition formula UNSAT, done"); (true, depth, formula, qrfs)
+
+(** This is a utility function that creates the delta symbols for each variable symbol,
+    and relates the delta symbols with the original variable symbols by saying
+    that each dx = x' - x.
+*)
+let add_diff_terms_to_formula srk f x_xp =
+  List.fold_right
+    (fun (x, xp) (f, dx_list, dx_sym_set, x_to_dx, dx_to_x) -> 
+       let dname = String.concat "" ["d_"; show_symbol srk x] in
+       let cx = mk_const srk x in
+       let cxp = mk_const srk xp in 
+       let diff = mk_sub srk cxp cx in
+       let dx_sym = mk_symbol srk ~name:dname `TyInt in
+       let dx = mk_const srk dx_sym in
+       let f_with_dx = mk_and srk [f ; mk_eq srk dx diff] in
+       (f_with_dx, 
+        List.cons dx_sym dx_list,
+        Symbol.Set.add dx_sym dx_sym_set, 
+        Symbol.Map.add x dx_sym x_to_dx, 
+        Symbol.Map.add dx_sym x dx_to_x)
+    )
+    x_xp
+    (f, [], Symbol.Set.empty, Symbol.Map.empty, Symbol.Map.empty)
+
+(* let simplify_residual_formula srk x_set xp_set residual_formula = 
+  let polka = Polka.manager_alloc_strict () in
+  let exists = fun x -> (Symbol.Set.mem x x_set || Symbol.Set.mem x xp_set) in
+  Abstract.abstract ~exists:exists srk polka residual_formula
+  |> SrkApron.formula_of_property *)
+(** The actual swf operator only has true or false as outcomes, corresponding to
+    able to prove or unable to prove results given here.
+ *)
+let compute_swf srk tf =
```

Comment:
Rename.  The formula that's returned is not a sufficient well-foundedness condition (it's a residual), and we should try our best to match the vocabulary in the paper.